### PR TITLE
[GPII-3507]: Monitor CouchDB snapshots with log-based metrics

### PR DESCRIPTION
**Pull request by natarajaya**
This adds support of log-based metrics to our Stackdriver Ruby client, so we can create custom metric  counters for `snapshot_created` events and errors, and then alerting policy based on those metrics.

Known limitations / issues:
- Some times it takes really long time for Stackdriver to digest newly created log-based metrics and make them available for use by alerting policies. During that time it keeps dying when trying to apply alert policies with `The metric referenced by the provided filter is unknown`. I am not sure of the solution to this at the moment.
- Log-based metric that supposed to check for `k8s-snapshots` general errors is pretty dumb. Since `k8s-snapshots` does not separate logs by severity, and there are not much of error examples, I don't have better proposal atm.
- Alert policy that based on log-based metrics fails in case any of those two conditions has not met:
  1. Absense of `k8s-snapshots` errors.
  1. Count of `snapshot_created` events for CouchDB is greater than 2 for 10 minutes.
This is does not exactly match reality (lowest replica count for CouchDB is 2 for dev clusters, with deltas of `PT5M PT15M PT45M`, which means that at least 4 new snapshots must be created in 10 minutes), but helps avoid issues with check alignment. I personally feel like it is good enough for the purpose, but we can always tweak this further if needed.

### CODE REVIEW

File: gcp/modules/gcp-stackdriver-alerting/client.rb [hunk 1]
```diff
@@ -65,6 +68,46 @@ def debug_output(resource)
...
+    puts "[DEBUG] log_based_metric: " + log_based_metric.inspect if @debug_mode
+    stackdriver_log_based_metrics[log_based_metric.name] = log_based_metric
+  end
+
+  log_based_metrics.each do |log_based_metric|
+    if stackdriver_log_based_metrics[log_based_metric["name"]]
+      puts "Updating log-based metric \"#{log_based_metric["name"]}\"..."
+      metric = stackdriver_logging_client.metric log_based_metric["name"]
+      metric.filter = log_based_metric["filter"]
+      metric.save
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: Since there is only single meaningful field in Stackdriver metrics (filter), I think we can.

**Comment by natarajaya** [comment]
@mrtyler @stepanstipl Thanks for your reviews!

Summary of recent changes:
- To workaround that nasty situation when log-based metric created, but not yet available for use by alerting policies, I had to move log-based metric creation into new module `gcp-stackdriver-lbm` and apply it at the same time with `gcp-stackdriver-exclusion` and `gcp-stackdriver-export` modules.
- To avoid code duplication I decided to make our Stackdriver Ruby client more flexible to be called in the same way from new `gcp-stackdriver-lbm` module.
- For the lack of better place, I put Stackdriver client ruby file into `shared/rakefiles` dir.
- I changed alert policy configuration:
  1. *K8s-snapshots Errors Check* is triggered when `logging.googleapis.com/user/k8s_snapshots.error` rate is higher than 1 for 5 minutes, which seemed to me like reasonable way to get notified in case there are persistent errors in `k8s-snapshots` logs.
  1. *K8s-snapshots Check for CouchDB* is triggered when `logging.googleapis.com/user/k8s_snapshots.couchdb.snapshot_created` metric is absent for 5 minutes. Backup deltas for CouchDB starts with `PT5M` for all environments, meaning that there should be at least one successful backup event in 5 minutes, so this policy configuration seemed sufficient to trigger alerts in case snapshot creation events are missing in `k8s-snapshots` logs.
- `gcp-stackdriver-alerting` module renamed into `gcp-stackdriver-monitoring` to make more sense, since it is dealing with other resources except alerting policies.

There are still couple of issues:
- Issue with module dependencies: since `gcp-stackdriver-lbm` is now independent module, it would be handled one of the first on `rake destroy` as well, and would fail, trying to delete log-based metrics that are still in use by alerting policies of `gcp-stackdriver-monitoring` module. As workaround, I removed null resource that triggered on destroy from `gcp-stackdriver-lbm` module and now all log-based metrics are being deleted during `gcp-stackdriver-monitoring` module destruction.
- Issue with new log-based metrics and long-living environments: for `prd` and `stg`, there can still be a situation, when there is not enough time passing between creation of newly added log-based metric and application of new alert policy based on that metric. This will result in `The metric referenced by the provided filter is unknown` errors, fixable by retry. I think we can live with that after adding troubleshooting note describing this situation into README.

**Comment by mrtyler** [comment]
LGTM

Can we solve both ordering issues by making `monitoring` depend on `lbm` via terraform.tfvars?

**Comment by natarajaya** [comment]
@mrtyler Unfortunately, even after adding dependency on `gcp-stackdriver-lbm`, both modules are being destroyed nearly at the same time. I returned destroy null resource back into `gcp-stackdriver-lbm` and added sleep to prevent race condition with `gcp-stackdriver-monitoring` module. I think this looks better now.

File: gcp/modules/gcp-stackdriver-lbm/main.tf [hunk 2]
```diff
@@ -48,3 +48,40 @@ resource "null_resource" "apply_stackdriver_lbm" {
...
+resource "null_resource" "destroy_stackdriver_lbm" {
+  provisioner "local-exec" {
+    when = "destroy"
+
+    command = <<EOF
+      export PROJECT_ID=${var.project_id}
+      export GOOGLE_CLOUD_KEYFILE=${var.serviceaccount_key}
+
+      # To avoid race condition with gcp-stackdriver-monitoring module
+      sleep 30
```
Comment (mrtyler) [inline gcp/modules/gcp-stackdriver-lbm/main.tf]: How about a message here so the viewer knows this module isn't stuck?

Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-lbm/main.tf]: Done.

**Comment by natarajaya** [comment]
> :(. This sounds like a Terragrunt bug.

@mrtyler Nevermind, looks like I somehow did not add `lbm` dependency to `dev` and `prd` environments.

**Pull request merged**